File: 19-2.ex
#!/usr/bin/env elixir

defmodule Sim do

  def sim(index, limit, limit), do: {limit, limit - index}

  def sim(index, size, limit) do
    add = rem(index+div(size+1,2) - 1,  size) + 1
    size = size + 1
    index =
      if (index > add) do
        index
      else
        rem(size + index - 1, size)
      end
    sim(index, size, limit)
  end

  def calculate(input) do
    sim(0, 1, input)
  end
end

input = 3005290
result = Sim.calculate(input)
IO.inspect result

File: 19-2.py
#!/usr/bin/env python3

def next_state(index, size, limit):
    while True:
        if size == limit:
            return (limit, limit - index)

        add = ((index + ((size+1) // 2) - 1) % size ) + 1
        size = size + 1

        if index <= add:
           index = (size + index - 1) % size



i = 3005290
result = next_state(0,1,i)

print(result)

